bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaAccountHandler.java

```diff
@@ -0,0 +1,550 @@
+                }
+
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            } else if ((System.currentTimeMillis() - apiIntervalTimestamp) > 1000 * API_ERROR_INTERVAL_SECONDS) {
+                logger.trace("Resetting the error counter. ({} errors in the last interval)", apiIntervalErrors);
+                apiIntervalTimestamp = System.currentTimeMillis();
+                apiIntervalErrors = 0;
+            }
+        }
+
+        return false;
+    }
+
+    protected Vehicle[] queryVehicles() {
+
+        String authHeader = getAuthHeader();
+
+        if (authHeader != null) {
+            // get a list of vehicles
+            Response response = vehiclesTarget.request(MediaType.APPLICATION_JSON_TYPE)
+                    .header("Authorization", authHeader).get();
+
+            logger.debug("Querying the vehicle: Response: {}:{}", response.getStatus(), response.getStatusInfo());
+
+            if (!checkResponse(response, true)) {
+                logger.error("An error occurred while querying the vehicle");
+                return null;
+            }
+
+            JsonParser parser = new JsonParser();
+
+            JsonObject jsonObject = parser.parse(response.readEntity(String.class)).getAsJsonObject();
+            Vehicle[] vehicleArray = gson.fromJson(jsonObject.getAsJsonArray("response"), Vehicle[].class);
+
+            for (int i = 0; i < vehicleArray.length; i++) {
+                for (VehicleListener listener : vehicleListeners) {
+                    listener.vehicleFound(vehicleArray[i]);
+                }
+                for (Thing vehicle : getThing().getThings()) {
+                    if (vehicleArray[i].vin.equals(vehicle.getConfiguration().get(VIN))) {
+                        TeslaVehicleHandler vehicleHandler = (TeslaVehicleHandler) vehicle.getHandler();
+                        if (vehicleHandler != null) {
+                            logger.debug("Querying the vehicle: VIN {}", vehicleArray[i].vin);
+                            String vehicleJSON = gson.toJson(vehicleArray[i]);
+                            vehicleHandler.parseAndUpdate("queryVehicle", null, vehicleJSON);
+                            if (logger.isTraceEnabled()) {
+                                logger.trace("Vehicle is id {}/vehicle_id {}/tokens {}", vehicleArray[i].id,
+                                        vehicleArray[i].vehicle_id, vehicleArray[i].tokens);
+                            }
+                        }
+                    }
+                }
+            }
+            return vehicleArray;
+        } else {
+            return new Vehicle[0];
+        }
+    }
+
+    private String getStorageKey() {
+        return this.getThing().getUID().getId();
+    }
+
+    private ThingStatusInfo authenticate() {
+        TokenResponse token = logonToken;
+        SimpleDateFormat dateFormatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss");
```

Comment:
Better use [DateTimeFormatter](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html) which is both thread-safe and immutable. You also don't have to recreate it that often in that case.